HDDS-8800. Let `ozone sh key list` support listing all keys of a volume

ashishkumar50 opened the pull request:
## What changes were proposed in this pull request?

Currently there is no way to list all the keys inside volume. fs recursive list is limited to FSO/legacy bucket only. 
In this PR, now providing new shell command which can now list all the keys inside volume irrespective of bucket types.

- List keys inside volume, (by default max 100 keys are listed)

```
ozone sh key list o3://localhost:9862/vol1         
[ {
  "volumeName" : "vol1",
  "bucketName" : "bucket1",
  "name" : "key1",
  "dataSize" : 0,
  "creationTime" : "2023-06-12T11:20:51.325Z",
  "modificationTime" : "2023-06-12T11:20:51.378Z",
  "replicationConfig" : {
    "replicationFactor" : "THREE",
    "requiredNodes" : 3,
    "replicationType" : "RATIS"
  },
  "metadata" : { }
}, {
  "volumeName" : "vol1",
  "bucketName" : "bucket1",
  "name" : "key2",
  "dataSize" : 0,
  "creationTime" : "2023-06-12T11:20:57.385Z",
  "modificationTime" : "2023-06-12T11:20:57.414Z",
  "replicationConfig" : {
    "replicationFactor" : "THREE",
    "requiredNodes" : 3,
    "replicationType" : "RATIS"
  },
  "metadata" : { }
} ]
[ {
  "volumeName" : "vol1",
  "bucketName" : "bucket2",
  "name" : "key5",
  "dataSize" : 0,
  "creationTime" : "2023-06-12T11:21:10.722Z",
  "modificationTime" : "2023-06-12T11:21:10.750Z",
  "replicationConfig" : {
    "replicationFactor" : "THREE",
    "requiredNodes" : 3,
    "replicationType" : "RATIS"
  },
  "metadata" : { }
} ]
[ {
  "volumeName" : "vol1",
  "bucketName" : "bucket3",
  "name" : "key10",
  "dataSize" : 0,
  "creationTime" : "2023-06-12T11:21:36.127Z",
  "modificationTime" : "2023-06-12T11:21:36.157Z",
  "replicationConfig" : {
    "replicationFactor" : "THREE",
    "requiredNodes" : 3,
    "replicationType" : "RATIS"
  },
  "metadata" : { }
}, {
  "volumeName" : "vol1",
  "bucketName" : "bucket3",
  "name" : "obsbkt/key10",
  "dataSize" : 0,
  "creationTime" : "2023-06-12T11:22:04.486Z",
  "modificationTime" : "2023-06-12T11:22:04.515Z",
  "replicationConfig" : {
    "replicationFactor" : "THREE",
    "requiredNodes" : 3,
    "replicationType" : "RATIS"
  },
  "metadata" : { }
} ]
```

- List keys inside volume by giving -l option

```
ozone sh key list -l 2 o3://localhost:9862/vol1/
[ {
  "volumeName" : "vol1",
  "bucketName" : "bucket1",
  "name" : "key1",
  "dataSize" : 0,
  "creationTime" : "2023-06-12T11:20:51.325Z",
  "modificationTime" : "2023-06-12T11:20:51.378Z",
  "replicationConfig" : {
    "replicationFactor" : "THREE",
    "requiredNodes" : 3,
    "replicationType" : "RATIS"
  },
  "metadata" : { }
}, {
  "volumeName" : "vol1",
  "bucketName" : "bucket1",
  "name" : "key2",
  "dataSize" : 0,
  "creationTime" : "2023-06-12T11:20:57.385Z",
  "modificationTime" : "2023-06-12T11:20:57.414Z",
  "replicationConfig" : {
    "replicationFactor" : "THREE",
    "requiredNodes" : 3,
    "replicationType" : "RATIS"
  },
  "metadata" : { }
} ]
Listing first 2 entries of the result. Use --length (-l) to override max returned keys.
```
Since there are more than 2 keys in volume vol1, it also prompts user that it listed only first 2 entries and suggest to use -l option to get desired number of keys.


## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8800

## How was this patch tested?

Verified using new integration test.

ashishkumar50 commented:
@sadanand48 @sumitagrawl, Please help to review.

adoroszlai commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/volume/VolumeListKeysHandler.java:
```diff
@@ -0,0 +1,74 @@
+public class VolumeListKeysHandler extends VolumeHandler {
+
+  @CommandLine.Mixin
+  private ListOptions listOptions;
```
`ListOptions` supports `start` and `prefix`, but they are ignored for this command, which is confusing.  It may be better to use a simple option (copied from `ListOptions`) for limiting the number of keys in the output.

adoroszlai reviewed (commented):
Thanks @ashishkumar50 for working on this.

Would it make sense to add it as `ozone sh volume key list`?  We could add other volume-level operations for keys later (e.g. `key count` or `key stats` might be useful).

errose28 commented:
@ashishkumar50 Could you add the invocation of the command to the PR description? I think that would be the part people would have the most comments on.

I think it could make sense to do this using the existing `key` subcommand. Right now, keys in a bucket are listed with `ozone sh key list /volume/bucket`. We would also support `ozone sh key list /volume` to list all the keys in a volume.

ashishkumar50 commented:
> @ashishkumar50 Could you add the invocation of the command to the PR description? I think that would be the part people would have the most comments on.
> 
> I think it could make sense to do this using the existing `key` subcommand. Right now, keys in a bucket are listed with `ozone sh key list /volume/bucket`. We would also support `ozone sh key list /volume` to list all the keys in a volume.

@errose28, Thanks for the suggestion, I have updated in list key subcommand itself. I have updated PR description with invocation details.

adoroszlai reviewed (commented):
Thanks @ashishkumar50 for updating the patch.

ashishkumar50 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/OzoneAddress.java:
```diff
@@ -370,12 +378,6 @@ public void ensureSnapshotAddress()
-          "Volume name is missing.");
-    } else if (bucketName.length() == 0) {
-      throw new OzoneClientException(
-          "Bucket name is missing.");
```
Thanks, I have added new handler for volume or bucket specific.

ashishkumar50 commented:
I think we should apply start and prefix to keys, i.e. in this call:

@adoroszlai, I have added start and prefix, which will list keys based on this for each bucket.

krishnaasawa1 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/keys/ListKeyHandler.java:
```diff
@@ -49,6 +49,15 @@ public class ListKeyHandler extends BucketSnapshotHandler {
+      throws IOException {
     String volumeName = address.getVolumeName();
     String bucketName = address.getBucketName();
     String snapshotNameWithIndicator = address.getSnapshotNameWithIndicator();
```
By default will show keys within snapshots as well or skip that in case of recursive commands?

ashishkumar50 replied:
It will not list snapshots when listing at volume or bucket level.

adoroszlai commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/OzoneAddress.java:
```diff
@@ -435,4 +435,19 @@ public void print(PrintStream out) {
   }
+
+  public void ensureVolumeOrBucketAddress() throws OzoneClientException {
+    if (keyName.length() > 0) {
+      if (OmUtils.isBucketSnapshotIndicator(keyName)) {
+        // If snapshot, ensure snapshot URI
+        ensureSnapshotAddress();
+      }
```
I think we should reject key addresses here.

ashishkumar50 replied:
Rejected key address.

adoroszlai commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/common/VolumeBucketHandler.java:
```diff
@@ -0,0 +1,37 @@
+import picocli.CommandLine;
+
+/**
+ * Base class for either volume or bucket command handlers.
```
```suggestion
 * Base class for commands that accept both volume and bucket.
```

ashishkumar50 replied:
Updated

adoroszlai reviewed (commented):
Thanks @ashishkumar50 for updating the patch.

errose28 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/keys/ListKeyHandler.java:
```diff
@@ -94,4 +103,40 @@ protected void execute(OzoneClient client, OzoneAddress address)
+      // More keys were returned notify about max length
+      if (keyIterator.hasNext() || (bucketIterator.hasNext()
+          && maxKeyLimit <= 0)) {
+        out().println("Listing first " + totalKeys + " entries of the " +
+            "result. Use --length (-l) to override max returned keys.");
+        return;
+      }
+    }
+    if (isVerbose()) {
+      out().printf("Found : %d keys in volume : %s %n",
+          totalKeys, volumeName);
+    }
```
Looks like `-l` for the existing list command does not support `-1` as an override to list all keys. It should probably have that, but that can be done in a follow up PR. These two things combined mean the length message will always be there and jq will never work unless -1 is set to something like 100 billion.

adoroszlai replied:
Filed HDDS-8858 to avoid mixing these messages with the JSON output.  I was thinking about showing only in `--verbose` mode, but `stderr` seems a better idea.

errose28 replied:
The separate Jira sounds good. There are probably other places where we have this problem as well. I guess that makes it optional to fix in this PR or not.

adoroszlai replied:
> I will create another Jira and handle -1 to override list all keys.

How about adding a `--all` option in `ListOptions`?  If it is set, then `limit` could be set to `MAX_VALUE` or `-1` (whichever works) _internally_.  I think that would be more user-friendly than using a special value publicly.

ashishkumar50 replied:
`--all` option sounds good.

errose28 replied:
Sounds good to me as well.

adoroszlai commented:
Thanks @ashishkumar50 for the patch, @devmadhuu, @errose28, @krishnaasawa1 for the review.
